### ISSUE TITLE
The dequeue process of activation queue is inefficient

### ISSUE THREAD

**Issue by hwwhww** [ethereum/eth2.0-specs#1511]
(Updated the table)

### Issue
By observing Lighthouse testnet status, I just realized that the validators dequeue rate is `MIN_PER_EPOCH_CHURN_LIMIT/2` per epoch instead of `MIN_PER_EPOCH_CHURN_LIMIT` per epoch.

It's because the current `process_registry_updates` function first selected the eligible 4 validators sorted by `activation_eligibility_epoch`, and then update the `validator.activation_epoch` (if `validator.activation_epoch == FAR_FUTURE_EPOCH`). However, it may select **the same validators** as the previous round.

The current code:
```python
    # Queue validators eligible for activation and not dequeued for activation prior to finalized epoch
    activation_queue = sorted([
        index for index, validator in enumerate(state.validators)
        if validator.activation_eligibility_epoch != FAR_FUTURE_EPOCH
        and validator.activation_epoch >= compute_activation_exit_epoch(state.finalized_checkpoint.epoch)
    ], key=lambda index: state.validators[index].activation_eligibility_epoch)
    # Dequeued validators for activation up to churn limit (without resetting activation epoch)
    for index in activation_queue[:get_validator_churn_limit(state)]:
        validator = state.validators[index]
        if validator.activation_epoch == FAR_FUTURE_EPOCH:
            validator.activation_epoch = compute_activation_exit_epoch(get_current_epoch(state))
```

To make it easier to understand, here is the scenario:
1. At epoch `816`:
    * 16 validators get set activation eligibility. `validator.activation_eligibility_epoch` = `816`.
    * 4 validators get selected to `activation_queue`.
    * These validators got set `validator.activation_epoch` to `821`
2. At epoch `817`:
    * Since we sort the queue by `key=lambda index: state.validators[index].activation_eligibility_epoch`, the tiebreaker key may be validator index. So **the same** `4` validators get selected to `activation_queue`. (Their `activation_epoch = 821` is greater than or equal to `compute_activation_exit_epoch(finalized_epoch) = 820`, matching the condition). 
    * Since `validator.activation_epoch` != `FAR_FUTURE_EPOCH`, **no validators' `activation_epoch` would be changed**.

![image](https://user-images.githubusercontent.com/9263930/70531976-05d0cf80-1b91-11ea-8811-d1e996ff3858.png)

So with a bunch of validators in the waiting queue, more than half of the epochs (depending on finalization) couldn't dequeue any validator. That's very inefficient.

### Proposed solution
1. Should we set condition to `validator.activation_eligibility_epoch >= compute_activation_exit_epoch(state.finalized_checkpoint.epoch)` instead of `validator.activation_epoch >= compute_activation_exit_epoch(state.finalized_checkpoint.epoch)`?
2.  Check if `validator.activation_epoch == FAR_FUTURE_EPOCH` in the filter.

If so, it can be changed to:
```python
    # Queue validators eligible for activation and not dequeued for activation prior to finalized epoch
    activation_queue = sorted([
        index for index, validator in enumerate(state.validators)
        if validator.activation_epoch == FAR_FUTURE_EPOCH
        and validator.activation_eligibility_epoch >= compute_activation_exit_epoch(state.finalized_checkpoint.epoch)
    ], key=lambda index: state.validators[index].activation_eligibility_epoch)
    # Dequeued validators for activation up to churn limit
    for index in activation_queue[:get_validator_churn_limit(state)]:
        validator = state.validators[index]
        validator.activation_epoch = compute_activation_exit_epoch(get_current_epoch(state))
```

**Comment by ChihChengLiang** [comment]
Just find @protolambda's data provides a good visual explanation of this issue. The activation is always odd number epochs here, and the 8 possible activations at 822 and 824 are wasted.

```
        :            pub ℹ️   eff.bal.    balance [  eligible activation       exit withdrawal]
   16390: 94c15c7778881c ▶️ 3200000000 3201135906 [       816        821         ♾️         ♾️]
   16391: 8b4ea3a9698642 ▶️ 3200000000 3201156649 [       816        821         ♾️         ♾️]
   16392: ae0498b6842f29 ▶️ 3200000000 3201216850 [       816        821         ♾️         ♾️]
   16393: 96b206fdff20b2 ▶️ 3200000000 3201204010 [       816        821         ♾️         ♾️]
   16394: 90c2056a1a4df8 ▶️ 3200000000 3201152177 [       816        823         ♾️         ♾️]
   16395: aa966edb8b9a18 ▶️ 3200000000 3201203190 [       816        823         ♾️         ♾️]
   16396: 97c0df7204b5bf ▶️ 3200000000 3201194294 [       816        823         ♾️         ♾️]
   16397: a9ee6cb7d32f29 ▶️ 3200000000 3201163514 [       816        823         ♾️         ♾️]
   16398: b9644b5184ac5c ▶️ 3200000000 3201123925 [       816        825         ♾️         ♾️]
   16399: 8b615deee1037b ▶️ 3200000000 3201109221 [       816        825         ♾️         ♾️]
   16400: b841dddbe701d0 ▶️ 3200000000 3201110792 [       816        825         ♾️         ♾️]
   16401: adf980f244ae26 ▶️ 3200000000 3201127506 [       816        825         ♾️         ♾️]
```
https://gist.githubusercontent.com/protolambda/82cbd4e8816e5aa6e093e55b924efa1e/raw/632a1cde449b7cb586104fe1c03c4b06b51ccdbb/registry.txt

**Comment by djrtwo** [comment]
closed via #1514
